src/librustc_typeck/check/mod.rs

```diff
@@ -4856,10 +4856,6 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             // Ignore for now.
             hir::StmtKind::Item(_) => {}
             hir::StmtKind::Expr(ref expr) => {
-                // Check with expected type of `()`.
-                self.check_expr_has_type_or_error(&expr, self.tcx.mk_unit());
```

Comment:
(T-Lang: N.B. this removes the expectation that `{ expr } : ()` and so therefore:
```rust
{ core::default::Default::default() }
1;
```
may stop compiling (crater will find out).

We could add inference defaulting to `()` if that is desirable.